cuda_core/cuda/core/experimental/_kernel_arg_handler.pyx

```diff
@@ -212,7 +212,13 @@ cdef class ParamHolder:
         for i, arg in enumerate(kernel_args):
             if isinstance(arg, Buffer):
                 # we need the address of where the actual buffer address is stored
-                self.data_addresses[i] = <void*><intptr_t>(arg.handle.getPtr())
+                if isinstance(arg.handle, int):
```

Comment:
Can we stomach the cost of an `isinstance` check here? 

* One alternative is to use a `try..except`, where the `try` part is cheap and the `except` part is expensive.

* Another alternative, which will eliminate the need to make any changes to the kernel arg handling logic here:
  - introduce a new type `HostPtr` which wraps an integer representing a pointer, and exposes a `getPtr()` method to get it.
  -  Expand the return type of [`Buffer.handle`](https://github.com/NVIDIA/cuda-python/blob/24fde1755ad1efa6e7d73b2236dad9681fa0b727/cuda_core/cuda/core/experimental/_memory.py#L78) to `DevicePtrT | HostPtr`
  - Change `LegacyPinnedMemoryResource` to return a buffer whose handle is a `HostPtr`.

Reply:
It seems like a reasonable approach and I agree it would simplify the handling here. A couple of comments:

* Perhaps we should rename `DevicePointerT` to just `PointerT`? In the case of pinned memory for instance, it doesn't actually represent a device pointer AFAIU.
* If we use the protocol as written, then `Union[IsHandleT, int]` is equivalent to just `IsHandle` (`int` type implements `__int__`). The protocol would also allow types like `float` or `bool`.  
  - I feel like this discussion has been had before, but it might be worth considering a protocol with a `__cuda_handle__()` method or something, rather than `__int__()`